src/coprocessor/codec/chunk/chunk.rs

```diff
@@ -0,0 +1,199 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// FIXME(shirly): remove following later
+#![allow(dead_code)]
+use super::Result;
+use super::column::Column;
+use coprocessor::codec::Datum;
+use std::sync::Arc;
+use tipb::expression::FieldType;
+
+const CHUNK_INITIAL_CAPACITY: usize = 32;
+
+/// Chunk stores multiple rows of data in Apache Arrow format.
+/// See https://arrow.apache.org/docs/memory_layout.html
+/// Values are appended in compact format and can be directly accessed without decoding.
+/// When the chunk is done processing, we can reuse the allocated memory by resetting it.
+pub struct Chunk {
+    columns: Vec<Column>,
+    // num_virtual_rows indicates the number of virtual rows, which have zero columns.
+    // It is used only when this Chunk doesn't hold any data, i.e. "len(columns)==0".
+    num_virtual_rows: usize,
+}
+
+impl Chunk {
+    ///new_chunk creates a new chunk with field types.
+    pub fn new(tps: &[FieldType]) -> Chunk {
+        Chunk::new_with_capacity(tps, CHUNK_INITIAL_CAPACITY)
+    }
+
+    pub fn new_with_capacity(tps: &[FieldType], cap: usize) -> Chunk {
+        let mut columns = Vec::with_capacity(tps.len());
+        for tp in tps {
+            columns.push(Column::new(tp, cap));
+        }
+        Chunk {
+            columns,
+            num_virtual_rows: 0,
+        }
+    }
+
+    // reset resets the chunk, so the memory it allocated can be reused.
+    // Make sure all the data in the chunk is not used anymore before you reuse this chunk.
+    pub fn reset(&mut self) {
+        for column in &mut self.columns {
+            column.reset();
+        }
+    }
+
+    // num_cols returns the number of rows in the chunk.
+    pub fn num_cols(&self) -> usize {
+        self.columns.len()
+    }
+
+    // num_rows returns the number of rows in the chunk.
+    pub fn num_rows(&self) -> usize {
+        if self.columns.is_empty() {
+            0
+        } else {
+            self.columns[0].length()
+        }
+    }
+
+    pub fn append_datum(&mut self, col_idx: usize, v: &Datum) -> Result<()> {
+        self.columns[col_idx].append_datum(v)
+    }
+}
+
+struct ArcChunk {
+    chunk: Arc<Chunk>,
+}
+
+impl ArcChunk {
+    pub fn new(chunk: Chunk) -> ArcChunk {
+        ArcChunk {
+            chunk: Arc::new(chunk),
+        }
+    }
+
+    //get_row gets the Row in the chunk with the row index.
+    pub fn get_row(&self, idx: usize) -> Row {
+        Row::new(self.chunk.clone(), idx)
+    }
+    // begin returns the first valid Row in the Chunk.
+    pub fn begin(&self) -> Row {
+        Row::new(self.chunk.clone(), 0)
+    }
+
+    //end returns a Row referring to the past-the-end element in the Chunk.
+    pub fn end(&self) -> Row {
+        let num_rows = self.chunk.num_rows();
+        Row::new(self.chunk.clone(), num_rows)
+    }
+}
+
+pub struct Row {
+    c: Arc<Chunk>,
+    idx: usize,
+}
+
+impl Row {
```

Comment:
Since we only planned to encode the return data to `Chunk`, maybe we don't need to implement `Row` and the data getters in `Column`?

Reply:
The implement of `Row` and `getXXX` are all used for the test now. And it's also used to remind the user how to use it. So could we keep it? Maybe we will need it in the future.